src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -1249,11 +1302,36 @@ func (s *service) SetWriteNewSeriesLimitPerShardPerSecond(
 	return s.GetWriteNewSeriesLimitPerShardPerSecond(ctx)
 }
 
-func (s *service) isOverloaded() bool {
-	// NB(xichen): for now we only use the database load to determine
-	// whether the server is overloaded. In the future we may also take
-	// into account other metrics such as CPU load, disk I/O rate, etc.
-	return s.db.IsOverloaded()
+func (s *service) SetDatabase(db storage.Database) error {
+	if atomic.LoadUint64(&s.dbIsSet) == 1 {
+		// Technically this check is racey in that it may not necessarily
+		// catch multiple calls to SetDatabase() if they happen concurrently,
+		// but it provides some basic sanity test against this function being
+		// called more than one time.
+		return errDatabaseHasAlreadyBeenSet
+	}
+
+	if db != nil {
+		s.db = db
+		s.nowFn = db.Options().ClockOptions().NowFn()
+		s.pools.id = db.Options().IdentifierPool()
+		atomic.StoreUint64(&s.dbIsSet, 1)
```

Comment:
This is racey, you need to do a compare and swap or else two goroutines could call this method at the same time and succeed.

Instead can you use the mutex already on the DB as per previous comment?  RLock should be nanoseconds.  

If you still would prefer an atomic, for the "checking" part, you can use the mutex to decide when to set the DB and then at the end of the SetDatabase method set a different atomic.  That way you avoid RLock on the read case, but still use the mutex for guarantees around setting the DB just once without having to use compare and set (i.e. prefer idiomatic vs specialized when idiomatic is just as good as specialized).